image-hd.c

```diff
@@ -166,6 +199,160 @@ static int hdimage_insert_ebr(struct image *image, struct partition *part)
+	ret = hdimage_insert_mbr(image, &mbr_list);
+	if (ret) {
+		image_error(image,"failed to write protective MBR\n");
+		return ret;
+	}
+
+	return 0;
+}
+
+static int hdimage_insert_gpt(struct image *image, struct list_head *partitions)
+{
+	struct hdimage *hd = image->handler_priv;
+	const char *outfile = imageoutfile(image);
+	struct gpt_header header;
+	struct gpt_partition_entry table[GPT_ENTRIES];
+	struct partition *part;
+	unsigned i, j;
+	int ret;
+
+	image_info(image, "writing GPT\n");
+
+	memset(&header, 0, sizeof(struct gpt_header));
+	memcpy(header.signature, "EFI PART", 8);
+	header.revision = htole32(GPT_REVISION_1_0);
+	header.header_size = htole32(sizeof(struct gpt_header));
+	header.current_lba = htole64(1);
+	header.backup_lba = htole64(image->size/512 - 1);
+	header.last_usable_lba = htole64(image->size/512 - 1 - GPT_SECTORS);
+	if (hd->disk_uuid) {
+		if (uuid_parse(hd->disk_uuid, header.disk_uuid) == -1) {
+			image_error(image, "invalid disk UUID: %s\n", hd->disk_uuid);
+			return -EINVAL;
+		}
+	}
+	else {
+		uuid_random(header.disk_uuid);
+	}
+	header.starting_lba = htole64(2);
+	header.number_entries = htole32(GPT_ENTRIES);
+	header.entry_size = htole32(sizeof(struct gpt_partition_entry));
+
+	i = 0;
+	memset(&table, 0, sizeof(table));
+	list_for_each_entry(part, partitions, list) {
+		if (header.first_usable_lba == 0)
+			header.first_usable_lba = htole64(part->offset / 512);
+
+		if (!part->in_partition_table)
+			continue;
+
+		if (strlen(part->partition_type_uuid) == 1)
+			part->partition_type_uuid =
+				gpt_partition_type_lookup(part->partition_type_uuid);
+		if (uuid_parse(part->partition_type_uuid, table[i].type_uuid) == -1) {
+			image_error(image,
+				    "part %s has invalid partition type UUID: %s\n",
+				    part->name, part->partition_type_uuid);
+			return -EINVAL;
+		}
+		if (part->partition_uuid) {
+			if (uuid_parse(part->partition_uuid, table[i].uuid) == -1) {
+				image_error(image,
+					    "part %s has invalid partition UUID: %s\n",
+					    part->name, part->partition_uuid);
+				return -EINVAL;
+			}
+		}
+		else {
+			uuid_random(table[i].uuid);
+		}
+		table[i].first_lba = htole64(part->offset/512);
+		table[i].last_lba = htole64((part->offset + part->size)/512 - 1);
+		for (j = 0; j < strlen(part->name) && j < 36; j++)
+			table[i].name[j] = htole16(part->name[j]);
+		i++;
+	}
+	header.table_crc = htole32(crc32(table, sizeof(table)));
+
+	header.header_crc = htole32(crc32(&header, sizeof(header)));
+	ret = insert_data(image, (char *)&header, outfile, sizeof(header), 512);
+	if (ret) {
+		image_error(image, "failed to write GPT\n");
+		return ret;
+	}
+	ret = insert_data(image, (char *)&table, outfile, sizeof(table), 2*512);
+	if (ret) {
+		image_error(image, "failed to write GPT table\n");
+		return ret;
+	}
+
+	ret = pad_file(image, NULL, outfile, image->size, 0x0, MODE_APPEND);
+	if (ret) {
+		image_error(image,
+			    "failed to write image partition '%s'\n",
+			    part->name);
```

Comment:
This error message makes no sense. There is no partition here. We're expanding the image to the full size here, right?